### PR TITLE
[MLIR][GPU-LLVM] Add GPU to LLVM-SPV address space mapping

### PR DESCRIPTION
**Pull request by victor-eds**
Implement mapping:

- `global`: 1
- `workgroup`: 3
- `private`: 0

Add `addressSpaceToStorageClass`, mapping GPU address spaces to SPIR-V storage classes to be able to use SPIR-V's
`storageClassToAddressSpace`, mapping SPIR-V storage classes to LLVM address spaces according to our mapping above *by definition*.

### CODE REVIEW

**Review by kuhar — APPROVED** [review]

File: mlir/include/mlir/Conversion/GPUCommon/AttrToSPIRVConverter.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#ifndef MLIR_CONVERSION_GPUCOMMON_ATTRTOSPIRVCONVERTER_H_
+#define MLIR_CONVERSION_GPUCOMMON_ATTRTOSPIRVCONVERTER_H_
+
+#include <mlir/Dialect/GPU/IR/GPUDialect.h>
+#include <mlir/Dialect/SPIRV/IR/SPIRVEnums.h>
```
Comment (FMarno) [inline mlir/include/mlir/Conversion/GPUCommon/AttrToSPIRVConverter.h:12]: Can either of these be replaced with a forward declaration?

File: mlir/test/Conversion/GPUToLLVMSPV/gpu-to-llvm-spv.mlir [hunk 2]
```diff
@@ -503,3 +503,15 @@ gpu.module @kernels {
...
 }
+
+// -----
+
+gpu.module @kernels {
+// CHECK-LABEL:   llvm.func spir_funccc @address_spaces(
+// CHECK-SAME:                                          {{.*}}: !llvm.ptr<1>
+// CHECK-SAME:                                          {{.*}}: !llvm.ptr<3>
+// CHECK-SAME:                                          {{.*}}: !llvm.ptr
+  gpu.func @address_spaces(%arg0: memref<f32, #gpu.address_space<global>>, %arg1: memref<f32, #gpu.address_space<workgroup>>, %arg2: memref<f32, #gpu.address_space<private>>) {
```
Comment (FMarno) [inline mlir/test/Conversion/GPUToLLVMSPV/gpu-to-llvm-spv.mlir:514]: can we get a default i.e. `memref<f32>`.

File: mlir/lib/Conversion/GPUCommon/AttrToSPIRVConverter.cpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <mlir/Conversion/GPUCommon/AttrToSPIRVConverter.h>
+
+namespace mlir {
+spirv::StorageClass addressSpaceToStorageClass(gpu::AddressSpace addressSpace) {
```
Comment (FMarno) [inline mlir/lib/Conversion/GPUCommon/AttrToSPIRVConverter.cpp:12]: constexpr?

File: mlir/test/Conversion/GPUToLLVMSPV/gpu-to-llvm-spv.mlir (hunk 2 above)
Comment (victor-eds) [reply mlir/test/Conversion/GPUToLLVMSPV/gpu-to-llvm-spv.mlir:514]: That maps to `0` right now. As we're just introducing `gpu.address_space` remapping, I'd argue that's out of scope for our use case here. We can discuss if we want to keep that in the future.

**Review by FMarno — APPROVED** [review]

**Pull request merged**